### PR TITLE
Add keyboard leftover30

### PR DESCRIPTION
**Pull request by marksard**
## Description

Add new keyboard -> leftover30

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/marksard/leftover30/config.h [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+/*== customize breathing effect ==*/
+  /*==== (DEFAULT) use fixed table instead of exp() and sin() ====*/
+  #define RGBLIGHT_BREATHE_TABLE_SIZE 256      // 256(default) or 128 or 64
+  /*==== use exp() and sin() ====*/
+  #define RGBLIGHT_EFFECT_BREATHE_CENTER 1.85  // 1 to 2.7
+  #define RGBLIGHT_EFFECT_BREATHE_MAX    255   // 0 to 255
+#endif
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCING_DELAY 5
```
Comment (drashna) [inline keyboards/marksard/leftover30/config.h:94]:
```suggestion
#define DEBOUNCE 5
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#ifdef RGBLIGHT_ENABLE
+//Following line allows macro to read current RGB settings
+extern rgblight_config_t rgblight_config;
+#endif
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:21]: This shouldn't be needed. Query functions for all of this information exist, now. 
```suggestion
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  _LOWER,
+  _RAISE,
+  _ADJUST,
+};
+
+enum custom_keycodes {
+  RGBRST = SAFE_RANGE,
+  LOWER,
+  RAISE,
+  KANJI,
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:34]: Formatting nit-pick. Also enums start at 0, no need to specify that.
```suggestion
    _BASE,
    _LOWER,
    _RAISE,
    _ADJUST,
};

enum custom_keycodes {
    RGBRST = SAFE_RANGE,
    LOWER,
    RAISE,
    KANJI,
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        unregister_code(KC_LANG2);
+      }
+      break;
+#ifdef RGBLIGHT_ENABLE
+    //led operations - RGB mode change now updates the RGB_current_mode to allow the right RGB mode to be set after reactive keys are released
+    case RGB_MOD:
+        if (record->event.pressed) {
+        rgblight_mode(RGB_current_mode);
+        rgblight_step();
+        RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:139]:
```suggestion
        RGB_current_mode = rgblight_get_mode();
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        }
+    break;
+    case RGBRST:
+        if (record->event.pressed) {
+        eeconfig_update_rgblight_default();
+        rgblight_enable();
+        RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:146]:
```suggestion
        RGB_current_mode = rgblight_get_mode();
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        }
+      } else {
+        unregister_code(KC_LANG2);
+      }
+      break;
+#ifdef RGBLIGHT_ENABLE
+    //led operations - RGB mode change now updates the RGB_current_mode to allow the right RGB mode to be set after reactive keys are released
+    case RGB_MOD:
+        if (record->event.pressed) {
+        rgblight_mode(RGB_current_mode);
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:137]: should use no-eeprom version here, that way, it doesn't write to it twice.
```suggestion
        rgblight_mode_noeeprom(RGB_current_mode);
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        else {
+          tap_code((clockwise == true) ? KC_WH_D : KC_WH_U);
+        }
+
+    }
+}
+
+void keyboard_post_init_user(void) {
+  #ifdef RGBLIGHT_ENABLE
+    RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:192]:
```suggestion
    RGB_current_mode = rgblight_get_mode();
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        else if (IS_LAYER_ON(_LOWER)) {
+          if (clockwise) {
+            SEND_STRING(SS_LCTL("y"));
+          } else {
+            SEND_STRING(SS_LCTL("z"));
+          }
+        }
+        else if (IS_LAYER_ON(_RAISE)) {
+          if (clockwise) {
+            SEND_STRING(SS_DOWN(X_LSFT)SS_TAP(X_DOWN)SS_UP(X_LSFT));
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:178]: This has a smaller compile size. 
```suggestion
            tap_code16(S(KC_DOWN));
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+          } else {
+            SEND_STRING(SS_DOWN(X_LSFT)SS_TAP(X_UP)SS_UP(X_LSFT));
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:180]:
```suggestion
            tap_code16(S(KC_UP));
```

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+
+  bool result = false;
+  switch (keycode) {
+    case KANJI:
+      if (record->event.pressed) {
+        if (keymap_config.swap_lalt_lgui == false) {
+          register_code(KC_LANG2);
+        } else {
+          SEND_STRING(SS_LALT("`"));
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:127]:
```suggestion
          tap_code16(A(KC_GRV));
```

or even:
```suggestion
          register_code16(A(KC_GRV));
```
and unregister of the same.

File: keyboards/marksard/leftover30/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+          else
+          {
+              rgblight_decrease_hue_noeeprom();
+          }
+        }
+        else if (IS_LAYER_ON(_LOWER)) {
+          if (clockwise) {
+            SEND_STRING(SS_LCTL("y"));
+          } else {
+            SEND_STRING(SS_LCTL("z"));
```
Comment (drashna) [inline keyboards/marksard/leftover30/keymaps/default/keymap.c:173]:
```suggestion
            tap_code16(LCTL(KC_Y));
          } else {
            tap_code16(LCTL(KC_Z));
```

File: keyboards/marksard/leftover30/leftover30.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+// These are only required if you want to perform custom actions.
+
+void keyboard_pre_init_user(void) {
+  /* Set CAPSLOCK indicator pin as output */
+  setPinOutput(D1);
+  /* Set NUMLOCK indicator pin as output */
+  setPinOutput(D2);
+}
+
+__attribute__((weak)) bool led_set_keymap(uint8_t usb_led) { return false; }
```
Comment (drashna) [inline keyboards/marksard/leftover30/leftover30.c:30]:
```suggestion
```

File: keyboards/marksard/leftover30/leftover30.c [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    writePinLow(D1);
+  }
+
+  if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
+    writePinHigh(D2);
+  }
+  else {
+    writePinLow(D2);
+  }
+}
```
Comment (drashna) [inline keyboards/marksard/leftover30/leftover30.c:49]:
```suggestion
bool led_update_kb(led_t led_state) {
    bool res = led_update_user(led_state);
    if(res) {
        writePin(D2, led_state.num_lock);
        writePin(D1, led_state.caps_lock);
    }
    return res;
}
```

**Comment by marksard** [comment]
drashna, thank you for your review!

File: keyboards/marksard/leftover30/config.h [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID 0xFEED
+#define PRODUCT_ID 0xDFA8
+#define DEVICE_VER 0x0001
+#define MANUFACTURER marksard
+#define PRODUCT leftover30
+#define DESCRIPTION A 30% mini keyboard
```
Comment (fauxpark) [inline keyboards/marksard/leftover30/config.h:28]:
```suggestion
```

File: keyboards/marksard/leftover30/config.h [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+#define LCD_DATA3_PIN    7            //< pin for 4bit data bit 3
+#define LCD_RS_PORT      LCD_PORT     //< port for RS line
+#define LCD_RS_PIN       3            //< pin  for RS line
+#define LCD_RW_PORT      LCD_PORT     //< port for RW line
+#define LCD_RW_PIN       2            //< pin  for RW line
+#define LCD_E_PORT       LCD_PORT     //< port for Enable line
+#define LCD_E_PIN        1            //< pin  for Enable line
+#endif
+*/
+
```
Comment (fauxpark) [inline keyboards/marksard/leftover30/config.h:253]:
```suggestion
```

File: keyboards/marksard/leftover30/keymaps/default/rules.mk [hunk 16]
```diff
@@ -0,0 +1,9 @@
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+#
+
+# If your custom treadstone32 pcb, you can rewrite to yes.
+RGBLIGHT_ENABLE = yes  # LED underglow (Enable WS2812 RGB underlight.)
+MOUSEKEY_ENABLE = yes       # Mouse keys
+ENCODER_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/marksard/leftover30/keymaps/default/rules.mk:9]:
```suggestion
RGBLIGHT_ENABLE = yes
MOUSEKEY_ENABLE = yes
ENCODER_ENABLE = yes
```

File: keyboards/marksard/leftover30/rules.mk [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# MCU name
+MCU = atmega32u4
+
+BOOTLOADER = caterina
```
Comment (fauxpark) [inline keyboards/marksard/leftover30/rules.mk:4]:
```suggestion
# Bootloader selection
BOOTLOADER = caterina
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by marksard** [comment]
fauxpark, thanks always.

**Review by fauxpark — APPROVED** [review]

File: keyboards/marksard/leftover30/readme.md [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+A 30% mini keyboard.
+
+* Keyboard Maintainer: [marksard](https://github.com/marksard)
+* Hardware Supported: The PCBs, controllers supported
+* Hardware Availability: Links to where you can find this hardware
+
+Make example for this keyboard (after setting up your build environment):
+
+    make leftover30:default
```
Comment (fauxpark) [inline keyboards/marksard/leftover30/readme.md:13]:
```suggestion
    make marksard/leftover30:default
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/marksard/leftover30/readme.md [hunk 19]
```diff
@@ -0,0 +1,17 @@
+# leftover30
+
+![leftover30](https://raw.githubusercontent.com/marksard/Keyboards/master/_image/_leftover30.jpg)
+
+A 30% mini keyboard.
+
+* Keyboard Maintainer: [marksard](https://github.com/marksard)
+* Hardware Supported: The PCBs, controllers supported
+* Hardware Availability: Links to where you can find this hardware
```
Comment (noroadsleft) [inline keyboards/marksard/leftover30/readme.md:9]: Please fill these lines out.

File: keyboards/marksard/leftover30/config.h [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+/*
+ * Split Keyboard specific options, make sure you have 'SPLIT_KEYBOARD = yes' in your rules.mk, and define SOFT_SERIAL_PIN.
+ */
+// #define SOFT_SERIAL_PIN D0  // or D1, D2, D3, E6
+
+// #define BACKLIGHT_PIN B7
+// #define BACKLIGHT_BREATHING
+// #define BACKLIGHT_LEVELS 3
+
```
Comment (noroadsleft) [inline keyboards/marksard/leftover30/config.h:63]:
```suggestion
```

File: keyboards/marksard/leftover30/config.h [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     B
+//#define MAGIC_KEY_BOOTLOADER_ALT ESC
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_EEPROM_CLEAR   BSPACE
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
+
```
Comment (noroadsleft) [inline keyboards/marksard/leftover30/config.h:181]:
```suggestion
```

File: keyboards/marksard/leftover30/config.h [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+
+/* disable these deprecated features by default */
+#define NO_ACTION_MACRO
+#define NO_ACTION_FUNCTION
+
+/* Bootmagic Lite key configuration */
+// #define BOOTMAGIC_LITE_ROW 0
+// #define BOOTMAGIC_LITE_COLUMN 0
```
Comment (noroadsleft) [inline keyboards/marksard/leftover30/config.h:204]:
```suggestion
```

File: keyboards/marksard/leftover30/keymaps/default/rules.mk [hunk 23]
```diff
@@ -0,0 +1,3 @@
+RGBLIGHT_ENABLE = yes
+MOUSEKEY_ENABLE = yes
```
Comment (noroadsleft) [inline keyboards/marksard/leftover30/keymaps/default/rules.mk:2]: I'm confused as to why these are defined here.

- Mouse keys is enabled at keyboard level, so setting it here is redundant
- RGB Light is disabled at keyboard level, only to enable it here

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/marksard/leftover30/keymaps/default/rules.mk (hunk 23 above)
Comment (marksard) [reply keyboards/marksard/leftover30/keymaps/default/rules.mk:2]: surely. Change the keyboard level.

File: keyboards/marksard/leftover30/readme.md (hunk 19 above)
Comment (marksard) [reply keyboards/marksard/leftover30/readme.md]: okay, thanks

**Review by noroadsleft — COMMENTED** [review]
I think I'd rather see `MOUSEKEY_ENABLE` and `RGBLIGHT_ENABLE` set at the keyboard level and not at the keymap level, but I'll let someone else make the final decision there.

**Comment by marksard** [comment]
> I think I'd rather see MOUSEKEY_ENABLE and RGBLIGHT_ENABLE set to yes at the keyboard level and not at the keymap level, but I'll let someone else make the final decision there.

I would like to provide MOUSEKEY_ENABLE and RGBLIGHT_ENABLE features in an accessible location so that users can turn them off at their discretion.
However, if this idea is not QMK's policy, I would specify it as the keyboard default.

**Comment by noroadsleft** [comment]
> I would like to provide MOUSEKEY_ENABLE and RGBLIGHT_ENABLE features in an accessible location so that users can turn them off at their discretion.
> However, if this idea is not QMK's policy, I would specify it as the keyboard default.

I'm thinking about QMK Configurator users mostly in this regard, because Configurator doesn't have any support for changing the compile settings, so users who use it to build their firmware won't be able to use Mousekeys or RGB Light, even if their hardware has all the necessary bits.

Turning the options off is as easy as adding `MOUSEKEY_ENABLE = no` or `RGBLIGHT_ENABLE = no` along side the `keymap.c`, and I'd prefer to make this easier for the people who don't know how to do that at the slight expense of those who are comfortable working with the code directly.

It's not an official policy, by any means.

**Comment by marksard** [comment]
I see. I think you have a point. Thanks for the advice.

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**